include/El/blas_like/level1/Copy/TranslateBetweenGrids.hpp

```diff
@@ -241,129 +247,4016 @@ void TranslateBetweenGrids
+            indexB = i;
+        }
     }
 
-    const Int rankA = A.RedundantRank();
-    const Int rankB = B.RedundantRank();
+    DistMatrix<T,STAR,VC,ELEMENT,D2>* B = dynamic_cast<DistMatrix<T,STAR,VC,ELEMENT,D2>*>(&(*B_Vector[indexB]));
 
-    simple_buffer<T,D1> sendBuffer(rankA == 0 ? height*width : 0);
-    simple_buffer<T,D2> bcastBuffer(B.Participating() ? height*width : 0);
+    const Int rowStrideB = B->RowStride();
+    const Int sizeB = B->Grid().VCSize();
+
+    const Int m = B->Height();
+    const Int n = B->Width();
+    const Int mLocB = B->LocalHeight();
+    const Int nLocB = B->LocalWidth();
+
+
+    const Int posInGrid = A.Grid().VCRank();
+    A.Resize(m,n); 
+    
+    Int rowStrideA = A.RowStride();
+    const Int sizeA = A.Grid().VCSize();
+    const Int rowGCD = GCD(rowStrideB, rowStrideA);
+    const Int rowLCM = rowStrideB*rowStrideA / rowGCD;
+
+
+    // Parent Subgrid Size: 4 Child Subgrid Size: 3
+    // Parent 0 1 2 3 0 1 2 3 0 1 2 3 
+    // Child  0 1 2 0 1 2 0 1 2 0 1 2
+
+    
 
+    const Int index_from = int(std::floor(posInGrid/sizeB));
     SyncInfo<D1> syncInfoA = SyncInfoFromMatrix(A.LockedMatrix());
-    SyncInfo<D2> syncInfoB = SyncInfoFromMatrix(B.LockedMatrix());
+    SyncInfo<D2> syncInfoB = SyncInfoFromMatrix(B->LockedMatrix());
 
-    // Send from the root of A to the root of B's matrix's grid
-    mpi::Request<T> sendRequest;
-    if(rankA == 0)
+
+    const bool inAGrid = A.Participating();
+    const bool inBGrid = indexB >=0 ? true:false;
+
+    const Int maxSendSize = mLocB * nLocB;
+
+    simple_buffer<T,D1> send_buf(maxSendSize, syncInfoB);
+    simple_buffer<T,D2> recv_buf(maxSendSize, syncInfoA);
+
+    T* sendBuf = send_buf.data();
+    T* recvBuf = recv_buf.data();
+
+    if(inBGrid)
     {
-        if (sendBuffer.size() != size_t(height*width))
-            RuntimeError("TranslateBetweenGrids: Bad sendBuffer size!");
+        copy::util::InterleaveMatrix(
+                mLocB, nLocB,
+                B->LockedBuffer(0,0),
+                1, B->LDim(),
+                sendBuf, 1, mLocB, syncGeneral);
 
-        util::InterleaveMatrix(
-            height, width,
-            A.LockedBuffer(), 1, A.LDim(),
-            sendBuffer.data(), 1, height, syncInfoA);
-        // TODO(poulson): Use mpi::Translate instead?
-        const Int recvRank = (usingViewingB ? B.Grid().VCToViewing(0) : 0);
-        mpi::ISend(
-            sendBuffer.data(), height*width, recvRank, activeCommB, sendRequest);
     }
+    //Synchronize(syncGeneral);
+    mpi::AllReduce( sendBuf, recvBuf, maxSendSize, mpi::SUM, allreduceComm,syncGeneral);
 
-    // Receive on the root of B's matrix's grid and then broadcast
-    // over the owning communicator
-    if(B.Participating())
+    if(inAGrid)
     {
-        if (bcastBuffer.size() != size_t(height*width))
-            RuntimeError("TranslateBetweenGrids: Bad bcastBuffer size!");
-        if(rankB == 0)
+        int sendWidth = int(n / rowLCM);
```

Comment:
Need to handle case where `n` doesn't cleanly divide `rowLCM`.

Reply:
Shall I update all the functions to handle this case or insert a check?

Reply:
I need to go over all the functions and understand their implementation if we want to support this case.